tests/integration/data/dns-provider-test/src/charm.py

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python3
+import logging
+
+from ops.charm import CharmBase
+from ops.framework import StoredState
+from ops.main import main
+from ops.model import ActiveStatus, WaitingStatus, ModelError
+
+logger = logging.getLogger(__name__)
+
+
+class DnsProviderTestCharm(CharmBase):
+    """Charm the service."""
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._stored.set_default(coredns_relation_data={})
+        self.is_related = False
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(
+            self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready
+        )
+        self.framework.observe(
+            self.on.dns_provider_relation_changed,
+            self._on_dns_provider_relation_changed,
+        )
+        self.framework.observe(
+            self.on.get_relation_data_action, self._on_get_relation_data_action
+        )
+
+    @property
+    def is_running(self):
+        try:
+            container = self.unit.get_container("httpbin")
+            return (
+                container.can_connect()
+                and container.get_service("httpbin").is_running()
+            )
+        except ModelError:
+            return False
+
+    def _on_install(self, event):
+        if not self.is_running:
+            self.unit.status = WaitingStatus("Waiting to start service")
+
+    def _on_httpbin_pebble_ready(self, event):
+        # Get a reference the container attribute on the PebbleReadyEvent
```

Comment:
In case I missed it, is the test charm just here for testing the relation with the supplied action?  If so -- why does it have to have a container running at all?

Reply:
It really doesnt. I shamelessly stole this from Corys istio charm  and just used the same setup he had (Im not sure his actually used the container either). This container is what gets populated when first calling charmcraft init, but yes its not technically used for anything currently. I suppose it does make things more "real"